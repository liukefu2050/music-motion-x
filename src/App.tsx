/* eslint-disable react/require-default-props */
/* eslint-disable react/forbid-prop-types */

import React, { Suspense } from "react"
import { StaticRouter, StaticRouterProps } from "react-router-dom/server"
import {
  QueryClient,
  QueryClientProvider,
  useQueryErrorResetBoundary,
} from "react-query"
import { BrowserRouter } from "react-router-dom"
import { ErrorBoundary } from "react-error-boundary"
import { Provider } from "react-redux"
import { RouteMatch, Routes, Route } from "react-router"
import ErrorFound from "@/components/ErrorPage"

import routes from "./routes"
import NotFound from "./pages/NotFound"
import PlayBar from "@/components/PlayBar"
import AppUpdateAvailable from "@/components/AppUpdateAvailable"

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      suspense: true,
    },
  },
})

const App = ({
  store,
  isServer,
  location,
  preloadedState,
}: {
  store: any
  isServer: boolean
  matchedRoutes?: RouteMatch[]
  location?: StaticRouterProps["location"]
  preloadedState: { [x: string]: any }
}) => {
  const { reset } = useQueryErrorResetBoundary()
  const content = (
    <>
      {/* <PageBack isBlack={false} /> */}
      <Suspense>
        <PlayBar />
      </Suspense>

      <Routes>
        {routes.map((r) => (
          <Route element={r.element} path={r.path} key={r.path}>
            {r.children &&
              r.children.length > 0 &&
              r.children.map((child) => (
                <Route
                  element={child.element}
                  path={child.path}
                  key={child.path}
                />
              ))}
          </Route>
        ))}
        <Route path="*" element={<NotFound />} />
      </Routes>
    </>
  )

  const IsomophicRouter = isServer ? (
    <StaticRouter location={location || ""}>{content}</StaticRouter>
  ) : (
    <BrowserRouter>
      <>
        <AppUpdateAvailable />
        {content}
      </>
    </BrowserRouter>
  )

  return (
    <ErrorBoundary
      onReset={reset}
      fallbackRender={({ resetErrorBoundary, error }) => (
        <ErrorFound resetErrorBoundary={resetErrorBoundary} error={error} />
      )}
    >
      <Provider store={store} serverState={preloadedState || {}}>
        <QueryClientProvider client={queryClient}>
          {IsomophicRouter}
        </QueryClientProvider>
      </Provider>
    </ErrorBoundary>
  )
}

export default App
