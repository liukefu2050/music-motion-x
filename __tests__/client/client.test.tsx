import App from "../../src/App"
import renderWithRoot from "../test-utils/renderWithModalRoot"
import getReduxStore from "../../src/store"
import { render, screen, waitFor } from "@testing-library/react"

describe("client ", () => {
  afterEach(() => {
    jest.useRealTimers()
  })
  it.skip("render CSR well", () => {
    jest.mock("../../src/routes", () => ({
      __esModule: true,
      default: [],
    }))
    const { asFragment } = renderWithRoot(
      <App store={getReduxStore({})} isServer={false} preloadedState={{}} />,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it.skip("render SSR well", async () => {
    jest.useFakeTimers()
    const { renderToString } = await import("react-dom/server")

    jest.mock("../../src/routes", () => ({
      __esModule: true,
      default: [],
    }))

    const container = document.createElement("div")
    container.id = "root"
    container.innerHTML = renderToString(
      <App
        store={getReduxStore({})}
        preloadedState={{}}
        location={"/"}
        isServer
        dehydratedState={undefined}
      />,
    )

    document.body.appendChild(container)

    const { asFragment } = render(
      <App
        store={getReduxStore({})}
        preloadedState={{}}
        location={"/"}
        isServer={false}
        dehydratedState={undefined}
      />,
      { container, hydrate: true },
    )

    await waitFor(() => screen.findByText("00:00"))

    expect(asFragment()).toMatchSnapshot()
  })
})
