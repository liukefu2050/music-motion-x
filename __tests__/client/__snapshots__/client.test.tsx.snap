// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client  render CSR well 1`] = `
<DocumentFragment>
  <audio
    preload="metadata"
    src=""
  />
  <div
    class="fixed bg-black transition-all overflow-hidden duration-300 left-4 bottom-5 z-[1000] w-32 h-10 rounded-[500px]"
    style="height: 40px;"
  >
    <div
      class=" flex items-center justify-center"
    >
      <div>
        <div
          class="play"
        />
      </div>
      <div
        class=" text-dg font-bold text-lg h-full leading-10 ml-3"
      >
        00:00
      </div>
    </div>
  </div>
  <div
    class=" w-16 h-16 z-[10000] relative my-0 mx-auto mb-0 "
    style="margin-top: 30vh;"
  >
    <div
      class="bounce w-full h-full rounded-full bg-secondary opacity-60 absolute top-0 left-0 animate-spin-loading"
    />
    <div
      class="bounce  w-full h-full rounded-full bg-secondary opacity-60 absolute top-0 left-0 animate-spin-loading"
      style="animation-delay: -1s;"
    />
  </div>
</DocumentFragment>
`;

exports[`client  render SSR well 1`] = `
<DocumentFragment>
  <audio
    preload="metadata"
    src=""
  />
  <div
    class="fixed bg-black transition-all overflow-hidden duration-300 left-4 bottom-5 z-[1000] w-32 h-10 rounded-[500px]"
    style="height: 40px;"
  >
    <div
      class=" flex items-center justify-center"
    >
      <div>
        <div
          class="play"
        />
      </div>
      <div
        class=" text-dg font-bold text-lg h-full leading-10 ml-3"
      >
        00:00
      </div>
    </div>
  </div>
  <!--$-->
  <main
    class=" p-4 relative overflow-hidden bg-mg"
  >
    <div
      class=" items-center bg-mg z-[99] block mt-5"
      style="transition: all 0.3s ease-in-out;"
    >
      <input
        class=" relative z-[1000] text-fg text-base border-none h-8 rounded-[200px] w-full bg- outline-none bg-black indent-8 transition-width duration-700 bg-[length:16px_16px] bg-no-repeat"
        placeholder=""
        style="background-image: url(test-file-stub);"
        type="text"
      />
      <span
        class=" ml-3 text-fg text-base whitespace-nowrap relative z-[1000]  invisible  w-0"
        style="transition: width 0.6s;"
      >
        取消
      </span>
    </div>
    <div
      class="block"
    />
  </main>
  <!--/$-->
</DocumentFragment>
`;
